extra/modules/fiftyone-devicedetection/src/main/java/org/prebid/server/hooks/modules/fiftyone/devicedetection/v1/hooks/FiftyOneDeviceDetectionRawAuctionRequestHook.java

```diff
@@ -0,0 +1,262 @@
+package org.prebid.server.hooks.modules.fiftyone.devicedetection.v1.hooks;
+
+import com.iab.openrtb.request.BidRequest;
+import com.iab.openrtb.request.BrandVersion;
+import com.iab.openrtb.request.Device;
+import com.iab.openrtb.request.UserAgent;
+import org.prebid.server.auction.model.AuctionContext;
+import org.prebid.server.hooks.execution.v1.auction.AuctionRequestPayloadImpl;
+import org.prebid.server.hooks.modules.fiftyone.devicedetection.core.detection.DeviceRefiner;
+import org.prebid.server.hooks.modules.fiftyone.devicedetection.core.mergers.MergingConfigurator;
+import org.prebid.server.hooks.modules.fiftyone.devicedetection.core.mergers.PropertyMerge;
+import org.prebid.server.hooks.modules.fiftyone.devicedetection.model.boundary.CollectedEvidence;
+import org.prebid.server.hooks.modules.fiftyone.devicedetection.model.boundary.CollectedEvidence.CollectedEvidenceBuilder;
+import org.prebid.server.hooks.modules.fiftyone.devicedetection.model.boundary.EnrichmentResult;
+import org.prebid.server.hooks.modules.fiftyone.devicedetection.model.config.AccountFilter;
+import org.prebid.server.hooks.modules.fiftyone.devicedetection.v1.adapters.DeviceMirror;
+import org.prebid.server.hooks.modules.fiftyone.devicedetection.v1.model.ModuleContext;
+import org.prebid.server.hooks.modules.fiftyone.devicedetection.v1.model.InvocationResultImpl;
+import org.prebid.server.hooks.v1.InvocationAction;
+import org.prebid.server.hooks.v1.InvocationResult;
+import org.prebid.server.hooks.v1.InvocationStatus;
+import org.prebid.server.hooks.v1.auction.AuctionInvocationContext;
+import org.prebid.server.hooks.v1.auction.AuctionRequestPayload;
+import org.prebid.server.hooks.v1.auction.RawAuctionRequestHook;
+import io.vertx.core.Future;
+import org.prebid.server.settings.model.Account;
+import org.prebid.server.util.ObjectUtil;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
+
+public class FiftyOneDeviceDetectionRawAuctionRequestHook implements RawAuctionRequestHook {
```

Comment:
That's my vision on the code. Please take as an example. And a few things that I often spot:
- according to out code style the closing figure `}` should be on the same line with the last method parameter (please use the same style from the example below)
- please get rid of pass builder as a parameter (often the object can be built just in place)
- I prefer not using PropertyMerge and MergingConfigurator, but instead doing merging in place (again, the example is below)
- please use such StringUtils, CollectionUtils and Optional to resolve cases when the object might be null
- get rid of protected methods that used only once in the same class (it's bad practice)

P.S. I left a few minor todos, you can ignore them if they don't resonate with you. Also I might have removed some quotes signs, looks like it's not needed (feel free to recover them if they really are needed)

```
public class FiftyOneDeviceDetectionRawAuctionRequestHook implements RawAuctionRequestHook {

    private static final String CODE = "fiftyone-devicedetection-raw-auction-request-hook";

    private final AccountFilter accountFilter;
    private final DeviceRefiner deviceRefiner;

    public FiftyOneDeviceDetectionRawAuctionRequestHook(AccountFilter accountFilter, DeviceRefiner deviceRefiner) {
        this.accountFilter = accountFilter;
        this.deviceRefiner = deviceRefiner;
    }

    @Override
    public String code() {
        return CODE;
    }

    @Override
    public Future<InvocationResult<AuctionRequestPayload>> call(AuctionRequestPayload payload,
                                                                AuctionInvocationContext invocationContext) {

        if (!isAccountAllowed(invocationContext)) {
            return Future.succeededFuture(
                    InvocationResultImpl.<AuctionRequestPayload>builder()
                            .status(InvocationStatus.success)
                            .action(InvocationAction.no_action)
                            .moduleContext(invocationContext.moduleContext())
                            .build());
        }

        //todo: bad naming doesn't correspond to the actual logic
        final ModuleContext moduleContext = addEvidenceToContext(
                (ModuleContext) invocationContext.moduleContext(),
                payload.bidRequest().getDevice());

        return  Future.succeededFuture(
                InvocationResultImpl.<AuctionRequestPayload>builder()
                        .status(InvocationStatus.success)
                        .action(InvocationAction.update)
                        .payloadUpdate(freshPayload -> updatePayload(freshPayload, moduleContext.collectedEvidence()))
                        .moduleContext(moduleContext)
                        .build());
    }

    private boolean isAccountAllowed(AuctionInvocationContext invocationContext) {
        if (accountFilter == null) {
            return true;
        }

        //todo: can the allowList be null?
        final List<String> allowList = accountFilter.getAllowList();
        return Optional.ofNullable(invocationContext)
                .map(AuctionInvocationContext::auctionContext)
                .map(AuctionContext::getAccount)
                .map(Account::getId)
                .filter(StringUtils::isNotBlank)
                .map(allowList::contains)
                .orElse(CollectionUtils.isEmpty(allowList));
    }

    private AuctionRequestPayload updatePayload(AuctionRequestPayload existingPayload,
                                                CollectedEvidence collectedEvidence) {

        final BidRequest currentRequest = existingPayload.bidRequest();

        final BidRequest patchedRequest = enrichDevice(currentRequest, collectedEvidence);
        if (patchedRequest == null || Objects.equals(patchedRequest, currentRequest)) {
            return existingPayload;
        }
        return AuctionRequestPayloadImpl.of(patchedRequest);
    }

    private BidRequest enrichDevice(BidRequest bidRequest, CollectedEvidence collectedEvidence) {
        if (bidRequest == null) {
            return null;
        }

        final Device enrichedDevice = ObjectUtils.defaultIfNull(bidRequest.getDevice(), Device.builder().build());

        final EnrichmentResult mergeResult = deviceRefiner.enrichDeviceInfo(
                new DeviceWrapper(enrichedDevice),
                modifyEvidenceWithDevice(collectedEvidence, bidRequest.getDevice()),
                DeviceWrapper.BUILDER_METHOD_SET.makeAdapter(enrichedDevice));

        if (mergeResult == null) {
            return null;
        }

        final Device mergedDevice = mergeResult.enrichedDevice();
        if (mergedDevice == null) {
            return null;
        }

        return bidRequest.toBuilder()
                .device(mergeResult.enrichedDevice())
                .build();
    }

    private ModuleContext addEvidenceToContext(ModuleContext moduleContext,
                                                 Device device) {

        final CollectedEvidence existingEvidence = Optional.ofNullable(moduleContext)
                .map(ModuleContext::collectedEvidence)
                .map(CollectedEvidence::toBuilder)
                .orElseGet(CollectedEvidence::builder)
                .build();

        return Optional.ofNullable(moduleContext)
                .map(ModuleContext::toBuilder)
                .orElseGet(ModuleContext::builder)
                .collectedEvidence(modifyEvidenceWithDevice(existingEvidence, device))
                .build();
    }

    private CollectedEvidence modifyEvidenceWithDevice(CollectedEvidence evidence, Device device) {
        if (device == null) {
            return evidence;
        }

        return evidence.toBuilder()
                .deviceUA(device.getUa())
                .secureHeaders(makeHeaders(device.getSua()))
                .build();
    }

    private Map<String, String> makeHeaders(UserAgent userAgent){
        final Map<String, String> headers = new HashMap<>();

        if (userAgent == null) {
           return headers;
        }

        if (CollectionUtils.isNotEmpty(userAgent.getBrowsers())) {
            final String fullUA = brandListToString(userAgent.getBrowsers());
            headers.put("header.Sec-CH-UA", fullUA);
            headers.put("header.Sec-CH-UA-Full-Version-List", fullUA);
        }

        headers.put("header.Sec-CH-UA-Mobile", "?" + userAgent.getMobile());

        //todo: can be re-written using Optional
        if (StringUtils.isNotEmpty(userAgent.getArchitecture())) {
            headers.put("header.Sec-CH-UA-Arch", toHeaderSafe(userAgent.getArchitecture()));
        }

        if (StringUtils.isNotEmpty(userAgent.getBitness())) {
            headers.put("header.Sec-CH-UA-Bitness", toHeaderSafe(userAgent.getBitness()));
        }

        if (StringUtils.isNotEmpty(userAgent.getModel())) {
            headers.put("header.Sec-CH-UA-Model", toHeaderSafe(userAgent.getModel()));
        }

        final BrandVersion platform = userAgent.getPlatform();
        if (StringUtils.isNotEmpty(platform.getBrand())) {
            headers.put("header.Sec-CH-UA-Platform", toHeaderSafe(platform.getBrand()));
        }

        if (CollectionUtils.isNotEmpty(platform.getVersion())) {
            headers.put("header.Sec-CH-UA-Platform-Version", String.join(".", platform.getVersion()));
        }

        return headers;

    }

    private static String brandListToString(List<BrandVersion> versions) {
        return versions.stream()
                .filter(version -> Objects.nonNull(version.getBrand()))
                .map(version -> toHeaderSafe(version.getBrand()) + ";v=" + String.join(".", version.getVersion()))
                .collect(Collectors.joining(", "));
    }

    private static String toHeaderSafe(String rawValue) {
        return rawValue.replace("\"", "\\\"");
    }
}
```